src/cordova/apple/OutlineAppleLib/Sources/OutlineTunnelSources/OutlineTunnel.swift

```diff
@@ -13,54 +13,156 @@
+        self.password = config["password"] as? String
+        self.method = config["method"] as? String
+        if let port = config["port"] {
+            self.port = String(describing: port)  // Handle numeric values
+        }
+        if let prefix = config["prefix"] as? String {
+            self.prefix = Data(prefix.utf16.map{UInt8($0)})
+        }
+    }
+    
+    public func encode() -> Data? {
+        return try? JSONEncoder().encode(self)
+    }
+    
+    public static func decode(_ jsonData: Data) -> OutlineTunnel? {
+        return try? JSONDecoder().decode(OutlineTunnel.self, from: jsonData)
+    }
+
+    // Helper function that we can call from Objective-C.
+    @objc public static func getTunnelNetworkSettings(tunnelRemoteAddress: String) -> NEPacketTunnelNetworkSettings {
+        // The remote address is not used for routing, but for display in Settings > VPN > Outline.
+        let settings = NEPacketTunnelNetworkSettings(tunnelRemoteAddress: tunnelRemoteAddress)
 
-  @objc
-  public enum TunnelStatus: Int {
-    case connected = 0
-    case disconnected = 1
-    case reconnecting = 2
-  }
+        // Configure VPN address and routing.
+        let vpnAddress = selectVpnAddress(interfaceAddresses: getNetworkInterfaceAddresses())
+        let ipv4Settings = NEIPv4Settings(addresses: [vpnAddress], subnetMasks: ["255.255.255.0"])
+        ipv4Settings.includedRoutes = [NEIPv4Route.default()]
+        ipv4Settings.excludedRoutes = getExcludedIpv4Routes()
+        settings.ipv4Settings = ipv4Settings
 
-  public convenience init(id: String, config: [String: Any]) {
-    self.init()
-    self.id = id
-    self.host = config["host"] as? String
-    self.password = config["password"] as? String
-    self.method = config["method"] as? String
-    if let port = config["port"] {
-      self.port = String(describing: port)  // Handle numeric values
+        // Configure with Cloudflare, Quad9, and OpenDNS resolver addresses.
+        settings.dnsSettings = NEDNSSettings(servers: ["1.1.1.1", "9.9.9.9", "208.67.222.222", "208.67.220.220"])
+        
+        return settings
     }
-    if let prefix = config["prefix"] as? String {
-      self.prefix = Data(prefix.utf16.map{UInt8($0)})
+}
+
+// Returns all IPv4 addresses of all interfaces.
+func getNetworkInterfaceAddresses() -> [String] {
+    var interfaces: UnsafeMutablePointer<ifaddrs>?
+    var addresses = [String]()
+    
+    if getifaddrs(&interfaces) != 0 {
+        NSLog("Failed to retrieve network interface addresses")
+        return addresses
     }
-  }
+    
+    var interface = interfaces
+    while interface != nil {
+        // Only consider IPv4 interfaces.
+        if interface?.pointee.ifa_addr.pointee.sa_family == UInt8(AF_INET) {
+            let addr = interface!.pointee.ifa_addr!.withMemoryRebound(to: sockaddr_in.self, capacity: 1) { $0.pointee.sin_addr }
+            if let ip = String(cString: inet_ntoa(addr), encoding: .ascii) {
```

Comment:
*nit* The original code used `stringWithUTF8String`, though no difference for an IP address.

```suggestion
            if let ip = String(cString: inet_ntoa(addr), encoding: .utf8) {
```